full-moon/src/ast/parsers.rs

```diff
@@ -1563,11 +1563,20 @@ cfg_if::cfg_if! {
         #[derive(Clone, Debug, PartialEq)]
         struct ParseSingleTypeInfo(TypeInfoContext);
         define_parser!(ParseSingleTypeInfo, TypeInfo, |this, state| {
-            let (state, base_type) = if let Ok((state, identifier)) = {
-                ParseIdentifier
+            // Singleton type info: `"yes" | "no"` in `(string) -> "yes" | "no"`, `"$$typeof"` in `{ ["$$typeof"]: number }`,
+            let (state, base_type) = if let Ok((state, string_singleton)) = ParseStringLiteral.parse(state) {
+                (state, TypeInfo::Basic(string_singleton))
+            // Singleton type info: `true` in `type X = Error & { handled: true }`
+            } else if let Ok((state, true_singleton)) = ParseSymbol(Symbol::True).parse(state) {
+                (state, TypeInfo::Basic(true_singleton))
+            } else if let Ok((state, false_singleton)) = ParseSymbol(Symbol::False).parse(state) {
+                (state, TypeInfo::Basic(false_singleton))
```

Comment:
Should these be changed to `TypeInfo::SingletonString` and `TypeInfo::SingletonBool` (or equivalent).

I don't know if there is any downstream consumer that does something like type inference, but this separation may make it easier to work with.

The Luau parser also does this (https://github.com/Roblox/luau/blob/4b96f7efc12f04fef2f9511811c807bb30dedaef/Ast/include/Luau/Ast.h#L1184-L1196) so it may be worthwhile following what they do

Reply:
I tried doing this locally, and it has ripple effects relating to StringLiteralQuoteType that appear to be far-reaching. I'm not against the idea, but would rather handle that in a separate PR.

@jeparlefrancais do you use full-moon for inference in darklua? Given the unique aspects of Luau's type system, especially in complex scenarios, means that reimplementing it in a way that doesn't trigger false positives or incorrect auto-completion seems slim.

Reply:
I meant for it to still be defined as `TypeInfo::String(TokenReference)` or `TypeInfo::Boolean(TokenReference)` - the same as how Value is defined for a String/Number/Boolean in an expression: https://github.com/Kampfkarren/full-moon/blob/7f6ae19a7d2eb2d0ab941c8712518d1885d1a71c/full-moon/src/ast/mod.rs#L333-L338


Although maybe it's just minor for no benefit? Could just pattern match on the TokenReference for TypeInfo::Basic

Reply:
@matthargett No unfortunately darklua has its own AST for mutations and manipulations

Reply:
I think since TypeInfo wasn't created in this PR, we can probably leave it like that for now. 

It could be refactored in another PR though, I think it's good to have unique variants because you pick a single one if you only need one. If an enum has a variant that is a group (like `Basic` or `Symbol`) then the pattern matching becomes a lot uglier if you need to match on the group first, and then match again on TokenReference to know what kind of token it.

It's simpler if it's all separated, because if you need multiple you can just match multiple into a single pattern branch.